client/src/router.ts

```diff
@@ -78,25 +72,51 @@ export const router = new Router({
     ],
 });
 
-router.beforeEach(async (to, from, next) => {
-    if (!coreStore.initialized && to.path !== "/_load") {
-        next({ path: "/_load" });
-        const response = await fetch("/api/auth");
-        if (response.ok) {
-            const data = await response.json();
-            if (data.auth) {
-                coreStore.setAuthenticated(true);
-                coreStore.setUsername(data.username);
-                coreStore.setEmail(data.email);
+router.beforeEach(async (to, _from, next) => {
+    coreStore.setLoading(true);
+    if (to.matched.some(record => record.meta.auth)) {
+        // authentication is required, but the store isnt authenticated
+        if (!coreStore.authenticated) {
+            console.log("Not authenticated");
+            // verify with the api to see if really authenticated
+            const response = await fetch("/api/auth");
+            if (response.ok) {
+                const data = await response.json();
+                if (data.auth) {
+                    // api reports authentication, adjust state and proceed
+                    coreStore.setAuthenticated(true);
+                    coreStore.setUsername(data.username);
+                    coreStore.setEmail(data.email);
+                    coreStore.setInitialized(true);
+                    next();
+                } else {
+                    // api reports not authenticated, redirect
+                    next({
+                        path: "/auth/login",
+                        query: {
+                            redirect: to.path,
+                        },
+                    });
+                }
+            } else {
+                // truly not authenticated, redirect to login
+                next({
+                    path: "/auth/login",
+                    query: {
+                        redirect: to.path,
+                    },
+                });
```

Comment:
I feel like this currently is just a more verbose way to write the original code.  What do you want to accomplish here that couldn't be accomplished with the old code and just dropping the `/_load` if case and next() call?

Reply:
So this is a little more complex, but I feel it gives a little more flexibility.

The old logic was making the API call on every directory wasn't _load if you weren't initialized. Now on a refresh event this will only make that call on a ```meta.auth``` path.

I tried a couple different things, actually tried to push all the API/authentication login within the coreStore but couldn't get things to wait properly for some variables in the router to populate to do it.

I also broke out my login redirect ```next()``` calls, which make it appear a little bigger. I could collapse those if you'd like.

Reply:
In practice the old code would only do the auth call on the very first route enter though.  As the store will be initialized after this first visit.

Reply:
Yeah exactly, all this really does different is initialize the store on the first time it's needed.

If you go line by line the new implementation is actually less lines if you format the next reroutes the same way, so I guess I'm confused where the verbosity is at.

Reply:
I must be missing something here then.  This is the old code without the _load, but using the new setLoading.  It's 22 lines 
```
router.beforeEach(async (to, from, next) => {
    coreStore.setLoading(true);
    if (!coreStore.initialized) {
        const response = await fetch("/api/auth");
        if (response.ok) {
            const data = await response.json();
            if (data.auth) {
                coreStore.setAuthenticated(true);
                coreStore.setUsername(data.username);
                coreStore.setEmail(data.email);
            }
            coreStore.setInitialized(true);
            router.push(to.path);
        } else {
            console.error("Authentication check could not be fulfilled.");
        }
    } else if (to.matched.some(record => record.meta.auth) && !coreStore.authenticated) {
        next({ path: "/auth/login", query: { redirect: to.path } });
    } else {
        next();
    }
});
```

and this is your suggestion with all comments stripped out and the next() arguments all on one line.  It's 26 lines and the second half contains a lot of } else {'s that require slightly more cognitive load to see what actually is happening.
```
router.beforeEach(async (to, _from, next) => {
    coreStore.setLoading(true);
    if (to.matched.some(record => record.meta.auth)) {
        if (!coreStore.authenticated) {
            const response = await fetch("/api/auth");
            if (response.ok) {
                const data = await response.json();
                if (data.auth) {
                    coreStore.setAuthenticated(true);
                    coreStore.setUsername(data.username);
                    coreStore.setEmail(data.email);
                    coreStore.setInitialized(true);
                    next();
                } else {
                    next({path: "/auth/login",query: {redirect: to.path,},});
                }
            } else {
                next({path: "/auth/login",query: {redirect: to.path,},});
            }
        } else {
            next();
        }
    } else {
        next();
    }
});
```

I also just feel that the original code is easier to follow.  First there is a check if the store has been initialized if not, perform the init steps.  After that there is a check to see if you're accessing a auth restricted page whether you are indeed authenticated.  and if the last else is reached you know everything has been resolved.

The current proposal also never sets the coreStore.initialized state until a auth route is visited. I'm aware there is nothing except the router that currently uses the initialized state, but it could very well be used for something in the future.

I hope I'm not coming over to aggressive here as I honestly appreciate the input, I'm just trying to get most out of it ;)

Reply:
What do you think of this?

```typescript
router.beforeEach(async (to, _from, next) => {
    coreStore.setLoading(true);
    if (to.matched.some(record => record.meta.auth)) {
        // authentication is required, but the store isnt authenticated
        if (!coreStore.initialized) {
            const response = await fetch("/api/auth");
            if (response.ok) {
                const data = await response.json();
                if (data.auth) {
                    coreStore.setAuthenticated(true);
                    coreStore.setUsername(data.username);
                    coreStore.setEmail(data.email);
                }
                coreStore.setInitialized(true);
                router.push(to.path);
            } else {
                console.error("Authentication check could not be fulfilled.");
            }
        } else {
            next();
        }
    } else {
        // no authentication required
        next();
    }
});
```

I get where you're coming from, the first one does appear simpler.

This takes your core initialization logic from the old code and removes any rerouting to _load. I think this is better than my previous one.

Just curious, if ```.itialized``` and ```.authenticated``` only ever get turned off/on in the same instances, what's the point of having both? You're not coming off aggressive, and I appreciate the input 8)

Reply:
I've got these changes in my latest commit, let me know what you think